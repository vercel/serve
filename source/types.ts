// source/types.ts
// Type definitions for the CLI.

// An error thrown by any native Node modules.
export declare interface NodeError extends Error {
  code: string;
}

// A path to a file/remote resource.
export declare type Path = string;
// The port to bind the server on.
export declare type Port = number;
// The name of the host.
export declare type Host = string;
// The address of the server.
export declare interface ServerAddress {
  local?: string;
  network?: string;
  previous?: number;
}

// The endpoint the server should listen on.
export declare type ListenEndpoint =
  | number
  | `tcp://${Host}:${Port}`
  | `unix:${Path}`
  | `pipe:\\\\.\\pipe\\${Host}`;

// The parsed endpoints.
export declare interface ParsedEndpoint {
  port?: Port;
  host?: Host;
}

// An entry for URL rewrites.
export declare interface Rewrite {
  source: string;
  destination: string;
}

// An entry for redirecting a URL.
export declare type Redirect = Rewrite & {
  type: number;
};

// An entry to send headers for.
export declare interface Header {
  source: string;
  headers: {
    key: string;
    value: string;
  }[];
}

// The configuration for the CLI.
export declare interface Configuration {
  public: Path;
  cleanUrls: boolean | Path[];
  rewrites: Rewrite[];
  redirects: Redirect[];
  headers: Header[];
  directoryListing: boolean | Path[];
  unlisted: Path[];
  trailingSlash: boolean;
  renderSingle: boolean;
  symlinks: boolean;
  etag: boolean;
}

// The options you can pass to the CLI.
export declare interface Options {
  '--help': boolean;
  '--version': boolean;
  '--listen': ParsedEndpoint[];
  '--single': boolean;
  '--debug': boolean;
  '--config': Path;
  '--no-request-logging': boolean;
  '--no-clipboard': boolean;
  '--no-compression': boolean;
  '--no-etag': boolean;
  '--symlinks': boolean;
  '--cors': boolean;
  '--no-port-switching': boolean;
  '--ssl-cert': Path;
  '--ssl-key': Path;
  '--ssl-pass': string;
}

// The arguments passed to the CLI (the options + the positional arguments)
export declare type Arguments = Partial<Options> & {
  _: string[];
};
