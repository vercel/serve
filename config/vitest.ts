// config/vitest.ts
// The vitest configuration file.

import { env } from 'node:process';
import { defineConfig } from 'vitest/config';

// Make sure the output of the CLI is in color, so that it matches the
// snapshots.
env.FORCE_COLOR = 2;

export default defineConfig({
  test: {
    // Collect coverage using C8.
    coverage: { enabled: true },
  },
});
