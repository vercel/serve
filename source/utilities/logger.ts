// source/utilities/logger.ts
// A simple colorized console logger.

/* eslint no-console: 0 */

import chalk from 'chalk';

const http = (...message: string[]) =>
  console.info(chalk.bgBlue.bold(' HTTP '), ...message);
const info = (...message: string[]) =>
  console.info(chalk.bgMagenta.bold(' INFO '), ...message);
const warn = (...message: string[]) =>
  console.error(chalk.bgYellow.bold(' WARN '), ...message);
const error = (...message: string[]) =>
  console.error(chalk.bgRed.bold(' ERROR '), ...message);
const log = console.log;

export const logger = { http, info, warn, error, log };
