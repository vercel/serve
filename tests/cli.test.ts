// tests/cli.test.ts
// Tests for the CLI part of the project.

import { env } from 'node:process';
import { afterEach, describe, test, expect, vi } from 'vitest';

import manifest from '../package.json';
import {
  getHelpText,
  parseEndpoint,
  checkForUpdates,
} from '../source/utilities/cli.js';
import { logger } from '../source/utilities/logger.js';
import { ParsedEndpoint } from '../source/types.js';

afterEach(() => {
  vi.restoreAllMocks();
});

// A list of cases used to test the `parseEndpoint` function. The first element
// is the name of the case, followed by the input to pass to the function,
// followed by the expected output.
type EndpointTestCase = [string, string, ParsedEndpoint];
const validEndpoints = [
  ['http port', '4242', { port: 4242 }],
  ['tcp url', 'tcp://localhost:4242', { port: 4242, host: 'localhost' }],
  ['unix socket', 'unix:///dev/sock1', { host: '/dev/sock1' }],
  ['pipe', 'pipe:\\\\.\\pipe\\localhost', { host: '\\\\.\\pipe\\localhost' }],
] as EndpointTestCase[];
// Another list of cases used to test the `parseEndpoint` function. The function
// should throw an error when parsing any of these cases, as they are invalid
// endpoints.
type InvalidEndpointTestCase = [string, string, RegExp];
const invalidEndpoints = [
  ['protocol', 'ws://localhost', /unknown.*endpoint.*scheme.*/i],
  ['unix socket', 'unix://', /invalid.*unix.*socket.*/i],
  ['windows pipe', 'pipe:\\localhost', /invalid.*pipe.*/i],
] as InvalidEndpointTestCase[];

describe('utilities/cli', () => {
  // Make sure the help message remains the same. If we are changing the help
  // message, then make sure to run `vitest` with the `--update-snapshot` flag.
  test('render help text', () => expect(getHelpText()).toMatchSnapshot());

  // Make sure the `parseEndpoint` function parses valid endpoints correctly.
  test.each(validEndpoints)(
    'parse %s as endpoint',
    (_name, endpoint, parsedEndpoint) =>
      expect(parseEndpoint(endpoint)).toEqual(parsedEndpoint),
  );

  // Make sure `parseEndpoint` throws errors on invalid endpoints.
  test.each(invalidEndpoints)(
    'parse %s as endpoint',
    (_name, endpoint, error) =>
      expect(() => parseEndpoint(endpoint)).toThrow(error),
  );

  // Make sure the update message is shown when the current version is not
  // the latest version.
  test('print update message when newer version exists', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    await checkForUpdates({
      ...manifest,
      version: '0.0.0',
    });

    expect(consoleSpy).toHaveBeenCalledOnce();
    expect(consoleSpy).toHaveBeenLastCalledWith(
      expect.stringContaining('UPDATE'),
      expect.stringContaining('latest'),
    );
  });

  // Make sure the update message is not shown when the latest version is
  // running.
  test('do not print update message when on latest version', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    await checkForUpdates({
      ...manifest,
      version: '99.99.99',
    });

    expect(consoleSpy).not.toHaveBeenCalled();
  });

  // Make sure an update check does not occur when the NO_UPDATE_CHECK env var
  // is set.
  test('do not check for updates when NO_UPDATE_CHECK is set', async () => {
    const consoleSpy = vi.spyOn(logger, 'log');

    env.NO_UPDATE_CHECK = 'true';
    await checkForUpdates({
      ...manifest,
      version: '0.0.0',
    });
    env.NO_UPDATE_CHECK = undefined;

    expect(consoleSpy).not.toHaveBeenCalled();
  });
});
