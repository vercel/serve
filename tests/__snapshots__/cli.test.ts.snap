// Vitest Snapshot v1

exports[`utilities/cli > render help text 1`] = `
"
  [1m[36mserve[39m[22m - Static file serving and directory listing

  [1mUSAGE[22m

    [1m$[22m [36mserve[39m --help
    [1m$[22m [36mserve[39m --version
    [1m$[22m [36mserve[39m folder_name
    [1m$[22m [36mserve[39m [-l [4mlisten_uri[24m [-l ...]] [[4mdirectory[24m]

    By default, [36mserve[39m will listen on [1m0.0.0.0:3000[22m and serve the
    current working directory on that address.

    Specifying a single [1m--listen[22m argument will overwrite the default, not supplement it.

  [1mOPTIONS[22m

    --help                              Shows this help message

    -v, --version                       Displays the current version of serve

    -l, --listen [4mlisten_uri[24m             Specify a URI endpoint on which to listen (see below) -
                                        more than one may be specified to listen in multiple places

    -p                                  Specify custom port

    -s, --single                        Rewrite all not-found requests to \`index.html\`

    -d, --debug                         Show debugging information

    -c, --config                        Specify custom path to \`serve.json\`

    -L, --no-request-logging            Do not log any request information to the console.

    -C, --cors                          Enable CORS, sets \`Access-Control-Allow-Origin\` to \`*\`

    -n, --no-clipboard                  Do not copy the local address to the clipboard

    -u, --no-compression                Do not compress files

    --no-etag                           Send \`Last-Modified\` header instead of \`ETag\`

    -S, --symlinks                      Resolve symlinks instead of showing 404 errors
    
    --ssl-cert                          Optional path to an SSL/TLS certificate to serve with HTTPS
                                        [90mSupported formats: PEM (default) and PKCS12 (PFX)[39m
    
    --ssl-key                           Optional path to the SSL/TLS certificate's private key
                                        [90mApplicable only for PEM certificates[39m

    --ssl-pass                          Optional path to the SSL/TLS certificate's passphrase

    --no-port-switching                 Do not open a port other than the one specified when it's taken.

  [1mENDPOINTS[22m

    Listen endpoints (specified by the [1m--listen[22m or [1m-l[22m options above) instruct [36mserve[39m
    to listen on one or more interfaces/ports, UNIX domain sockets, or Windows named pipes.

    For TCP ports on hostname \\"localhost\\":

      [1m$[22m [36mserve[39m -l [4m1234[24m

    For TCP (traditional host/port) endpoints:

      [1m$[22m [36mserve[39m -l tcp://[4mhostname[24m:[4m1234[24m

    For UNIX domain socket endpoints:

      [1m$[22m [36mserve[39m -l unix:[4m/path/to/socket.sock[24m

    For Windows named pipe endpoints:

      [1m$[22m [36mserve[39m -l pipe:\\\\\\\\.\\\\pipe\\\\[4mPipeName[24m
"
`;
